tools/devtools/src/devtools/contexts/YorkieSource.tsx

```diff
@@ -121,12 +140,69 @@ export function useYorkieDoc() {
   return value;
 }
 
+export enum TransactionEventType {
+  Document = 'document',
+  Presence = 'presence',
+}
+
+export const getTransactionEventType = (
+  event: TransactionEvent,
+): TransactionEventType => {
+  for (const docEvent of event) {
+    if (
+      docEvent.type === DocEventType.StatusChanged ||
+      docEvent.type === DocEventType.Snapshot ||
+      docEvent.type === DocEventType.LocalChange ||
+      docEvent.type === DocEventType.RemoteChange
+    ) {
+      return TransactionEventType.Document;
+    }
+  }
+
+  return TransactionEventType.Presence;
+};
+
 export function useTransactionEvents() {
-  const value = useContext(TransactionEventsContext);
-  if (value === undefined) {
+  const { events, hidePresenceEvents, setHidePresenceEvents } = useContext(
+    TransactionEventsContext,
+  );
+
+  if (events === undefined) {
     throw new Error(
       'useTransactionEvents should be used within YorkieSourceProvider',
     );
   }
-  return value;
+
+  // mark presence events as null in the original events
+  const presenceMarkedEvents = useMemo(() => {
+    if (!hidePresenceEvents) {
+      return events;
+    }
+
+    return events.map((event) => {
+      if (getTransactionEventType(event) === TransactionEventType.Presence) {
+        return null;
+      }
+
+      return event;
+    });
+  }, [events, hidePresenceEvents]);
+
+  // filter out presence events from the original events
+  const presenceFilteredEvents = useMemo(
+    () =>
+      events.filter(
+        (event) =>
+          getTransactionEventType(event) === TransactionEventType.Document,
+      ),
+    [hidePresenceEvents, events],
+  );
+
+  return {
+    originalEvents: events,
+    presenceMarkedEvents,
+    presenceFilteredEvents,
+    hidePresenceEvents,
+    setHidePresenceEvents,
+  };
```

Comment:
- Manage enhancedEvents (which adds metadata to events) and presenceFilteredEvents.

```js
  // create an enhanced events with metadata
  const enhancedEvents = useMemo(() => {
    return events.map((event) => {
      const transactionEventType = getTransactionEventType(event);
      return {
        event,
        transactionEventType,
        isFiltered:
          hidePresenceEvents &&
          transactionEventType === TransactionEventType.Presence,
      };
    });
  }, [hidePresenceEvents, events]);

  // filter out presence events from the original events
  const presenceFilteredEvents = useMemo(() => {
    if (!hidePresenceEvents) return enhancedEvents;
    return enhancedEvents.filter((e) => !e.isFiltered);
  }, [enhancedEvents]);

  return {
    originalEvents: enhancedEvents,
    presenceFilteredEvents,
    hidePresenceEvents,
    setHidePresenceEvents,
  };
```

Reply:
I think in this way code is much more clear and concise! I made the change :)